src/refresh/RefreshTimer.ts

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+/**
+ * The backoff time is between the minimum and maximum backoff time, based on the number of attempts.
+ * An exponential backoff strategy is used, with a jitter factor to prevent clients from retrying at the same time.
+ *
+ * The backoff time is calculated as follows:
+ *  - `basic backoff time` = `MinimumBackoffInMs` * 2 ^ `attempts`, and it is no larger than the `MaximumBackoffInMs`.
+ *  - based on jitter ratio, the jittered time is between [-1, 1) * `JitterRatio` * basic backoff time.
+ *  - the final backoff time is the basic backoff time plus the jittered time.
+ *
+ * Note: the backoff time usually is no larger than the refresh interval, which is specified by the user.
+ *  - If the interval is less than the minimum backoff, the interval is used.
+ *  - If the interval is between the minimum and maximum backoff, the interval is used as the maximum backoff.
+ *  - Because of the jitter, the maximum backoff time is actually `MaximumBackoffInMs` * (1 + `JitterRatio`).
+ */
+
+const MinimumBackoffInMs = 30 * 1000; // 30s
+const MaximumBackoffInMs = 10 * 60 * 1000; // 10min
+const MaxSafeExponential = 53; // Used to avoid overflow, as Number.MAX_SAFE_INTEGER = 2^53 - 1.
+const JitterRatio = 0.25;
+
+export class RefreshTimer {

```

Comment:
I think the class name RefreshTimer is misleading. I think backoffTimer is more appropriate.

Reply:
I agree, to some extent, if we want to re-use the timer elsewhere other than refresh. Then, the timer is neutral, and the refresh functionality uses a "backoffTimer" to calculate the backoff time. And we might also rename the method "canRefresh" back to "isBackoffEnd"... ;) 

@avanigupta @mrm9084 WDYT?

Reply:
`isBackoffDue` sounds better.

Reply:
When a refresh operation succeeds, the next server refresh will be triggered only after the specified refresh interval has expired. During this time, the provider cannot initiate any server operation - not because it's in backoff state, but because its refresh interval has not elapsed. Provider will be in backoff state only when the previous server refresh has failed. In general, "backoff" indicates that something bad has happened. That's why I'm hesitant to use the term "backoff" here. Backoff is our implementation detail.

Reply:
Ah yes, this whatever timer is coupled deeply with the refresh interval, and its purpose is to calculate the real interval after applying exponential backoff strategy. Then probably no need to reuse it, at least for the moment.